### ISSUE TITLE
Feature request: --ignore-tests <test id>

### ISSUE THREAD

**Issue by GottemHams** [drwetter/testssl.sh#1502]
Perhaps I'm overlooking something but I couldn't find anything in the manual nor the issues on here. Currently there doesn't seem to be a way to exclude a specific test when running with the default options, e.g. if you're aware of a certain risk so you want to ignore warnings about it.

__Example__
When you're using the default options it also runs the BREACH test. Now, you could just specify all other options yourself and simply exclude -B, but as tests get added/changed you'd continually have to keep the flags up to date. In the specific example of BREACH: I'm aware of the possible risk but personally I can live with this.

__Proposal__
Add an argument like `--ignore-tests <test id,test id 2,...>`, where the `test id` can be stuff like `BREACH,xxx,yyy,zzz`. These flags would correspond to the test IDs as they're shown in the output JSON: `"id": "BREACH"`. I'm assuming these don't usually change.

**Comment by drwetter** [comment]
> Perhaps I'm overlooking something 

Nope ;-) 

In general you can always pick the tests you like with the all switches and just omit what you don't like. That might be not as handy though as the suggestion above.

I was thinking about skipping tests too but what I had in mind then was just to use e.g. ``+B`` instead of ``-B`` - albeit the logic of ``+`` seems a bit upside down. Of course also your suggestion would work. (anybody else has a good proposal?)

Thx, Dirk

**Comment by GottemHams** [comment]
I only started using testssl a couple days ago so I'm not 100% familiar with all arguments yet. But I assume the following is what makes it test for "outdated" TLS versions (1.0, 1.1):
`-p, --protocols               checks TLS/SSL protocols (including SPDY/HTTP2)`

Now, while your suggestion of using `+B` to omit the BREACH test would also work, you can't skip __just__ the test for TLS 1.0. Like in case you __have__ to use 1.1 to support some older clients, but you don't want or need 1.0. However, by using exclusions based on test ID, you __can__ do that. Just exclude `TLS1` and you should be good, since `TLS1_1` is the ID for 1.1 and I'm assuming there are also `TLS1_2` and `TLS1_3`. The same would likely hold true for other "grouped" tests.

So `--ignore-tests` would be much more flexible and would have my preference.

**Comment by drwetter** [comment]
> .. you can't skip just the test for TLS 1.0. Like in case you have to use 1.1 to support some older clients, but you don't want or need 1.0. 

I kind of have to disappoint you here :-)  Depending on the circumstances: by the time you read it on the screen the check for the protocol version has been executed already for internal reasons. It saves time for later checks.

**Comment by GottemHams** [comment]
Perhaps that wasn't a very good example then, but I think you get the point. :D In the (special) case of TLS versions, you could simply omit the test from the output but still run it internally for proper functioning of the tool itself.

**Comment by FedericoCeratto** [comment]
In order to run automated tests in CI and spot regressions it would be nice to be able to tag known errors and display them as "known" or "ignored". The exit code should be unaffected by them, so that CI can be used to spot regressions.

Examples: ignore "Chain of trust NOT ok (self signed)" as CI uses self-signed certs.

**Comment by drwetter** [comment]
@GottemHams : yes, the point wasn't difficult to get. It isn't difficult to implement but I am scratching my head what a nice cmd line parameter then would look like. ``--no-breach`` would be kind of ok, ``-no-B`` does not sounds good as the short form gets long. OTOH  ``--breach=no`` or ``-B=no`` could work.

Opinions?

@FedericoCeratto : not sure I got what you meant.

**Comment by GottemHams** [comment]
Well, since the parameter can be reused/extended for other tests, something more generic should be better. I like having a fixed prefix such as `--no-` or perhaps `--skip-` to indicate a "subcategory", then just slam the test name behind it and presto. I did notice there's currently a flag `--nodns` though, not sure if that would cause confusion to some if you go with `--no-*`.

My second vote would be on the `--breach=no` format as well.

However, how would these work with excluding/skipping by internal test ID? Like what if I also wanted to exclude the `DNS_CAArecord` test? I don't see an argument that makes it check for this to begin with.

**Comment by drwetter** [comment]
> However, how would these work with excluding/skipping by internal test ID

this is the painful part. My previous comment was based on the existing _do*_  variables / corresponding  _run*_ functions. Those relate to existing cmd line switches.

Based on the JSON ids: that's is far off. E.g. ``run_server_defaults()`` has many JSON ids -- you can as well look at the screen output. A lot of the output is retrieved from the (or multiple) certificates, so it would require some effort to suppress output on our side - like give me the OCSP URI but not the CRL - when users can just throw away what is not needed.

**Comment by drwetter** [comment]
FYI: the framework for ``--no-*`` or ``--disable-*`` switch is now implemented in the branch _magnuslarsen-grading_dev_, see #1573 . Actually that was quite easy.

For JSON/CSV it'll be easier to filter on the user side.

**Comment by FedericoCeratto** [comment]
So far the only workaround that I found for detecting regressions in CI tests is:
```
testssl --ip=127.0.0.1 --nodns=none --csvfile testssl.tmp.csv --severity LOW https://127.0.0.1:8333 || true
sort testssl.tmp.csv > testssl.csv
diff testssl.csv tests/untestable/testssl_expected.csv
```
The sort is required because the order seemed to be non-deterministic. The last diff detects unexpected changes and set the exit code accordingly.

**Comment by drwetter** [comment]
Thanks for the hint, @FedericoCeratto . 

Sounds also a poor men's solution for #1085 .

> The sort is required because the order seemed to be non-deterministic

in what sense? Only if you're on the dev branch things change.
